Random changes

wm4 opened the pull request:
A bunch of unrelated minor improvements and cleanups.

wiiaboo commented on DOCS/man/vf.rst:
```diff
@@ -75,15 +85,20 @@ normal filter parameters.
-    negative numbers address the end of the list (-1 is the last).
+``--vf-del=filter``
+    Deletes the filter. The filter can even given the way it was added (filter
+    name and its full argument list), by lavel (prefixed with ``@``), or as
```
label*

kevmitch commented on common/codecs.h:
```diff
@@ -32,8 +31,8 @@ struct mp_decoder_list {
 
-void mp_add_decoder(struct mp_decoder_list *list, const char *family,
-                    const char *codec, const char *decoder, const char *desc);
+void mp_add_decoder(struct mp_decoder_list *list,  const char *codec,
```
extra space

kevmitch commented on filters/f_lavfi.c:
```diff
@@ -903,15 +903,58 @@ static bool is_usable(const AVFilter *filter, int media_type)
+    int offset= -1;
+    int count = 0;
+    for (const AVOption *o = av_opt_next(c, 0); o; o = av_opt_next(c, o)) {
+        // This is how libavfilter (at the time) decided to assing positional
```
assign

wm4 commented:
Updated; also added a bunch of commits on top of it.

Some should somewhat improve change behavior with filters like dynaudnorm etc. (when changing playback speed).

Others are for #5470, but probably no full solution.

sfan5 commented:
Can confirm that vapoursynth works properly now.

kevmitch commented on filters/filter.h:
```diff
@@ -47,6 +47,11 @@ bool mp_pin_in_write(struct mp_pin *p, struct mp_frame frame);
 // buffered in mp_pins goes against this.
 bool mp_pin_out_request_data(struct mp_pin *p);
 
+// As mp_pin_out_request_data(), but call the filter's process() function next
```
Same as?

kevmitch commented on filters/f_swresample.c:
```diff
@@ -450,15 +454,22 @@ static struct mp_frame filter_resample_output(struct priv *p,
+    // Limit the filtered data size for better latency when changing speed.
+    // Avoid buffering data within the resampler => restrict input size.
+    double s = p->opts->max_output_frame_size / 1000;
+    int max_in = MPCLAMP(s * p->in_rate/ p->out_rate * p->in_rate, 128, INT_MAX);
```
Since max_output_frame_size is time-based, doesn't it just need to be multiplied by the input rate to get the number of input frames for that duration? Or am I misunderstanding what this is calculating?

wm4 replied:
It includes the speed change, so if you e.g. play twice as fast, the input needs to be double the size to output a fragment of a certain actual duration. It tries to do that by using the sample rate ratio.

wm4 commented:
Rebased, split that one vapoursynth commit, added a bunch of new commits.

CC @occivink I think I implemented what you wanted for --lavfi-complex, but not sure.

wiiaboo commented:
Typo'd "lavfi" in 45036f906c1beb184fbb89fea9d5828ffe827fc6's title, btw.

haasn commented:
This patchset breaks building with `--enable-zsh-comp`

```
f543ddb8ef16a9713cfe0697badceb8e366c4e84 is the first bad commit
commit f543ddb8ef16a9713cfe0697badceb8e366c4e84
Author: wm4 <wm4@nowhere>
Date:   Wed Jan 31 06:50:21 2018 +0100

    options: slightly improve filter help output for lavfi bridge
    
    --vf=help will now list libavfilter filters, and e.g. --vf=yadif=help
    will list libavfilter filter options.
    
    The latter is rather bare, because the AVOption API is really awful
    (holy shit how is it so bad), and would require us to handle _every_
    option type manually.
    
    Alternatively we could call av_opt_show2(), which ffmpeg uses for help
    output in its CLI tools and which is much more detailed. But it's rather
    foreign and forces output through av_log(), so I don't really want to
    use it.
```

```
[236/236] Compiling TOOLS/zsh.pl
Use of uninitialized value $ao_str in scalar chomp at /home/nand/dev/mpvhq/TOOLS/zsh.pl line 46.

Waf: Leaving directory `/home/nand/dev/mpvhq/build'
Build failed
 -> task in '_mpv' failed (exit status 255): 
	{task 139853527539432: _mpv zsh.pl -> _mpv}
' "/usr/bin/perl" "/home/nand/dev/mpvhq/TOOLS/zsh.pl" "/home/nand/dev/mpvhq/build/mpv" > "etc/_mpv" '
```

haasn commented:
The culprit for the above seems to be the change to the format of `--ao=help`, which used to have a `:` separating the option name and the description, but now no longer does.

wm4 commented:
I don't care about the zsh and never did - not having to care was a promise made when the zsh script was accepted.
I'm in no hurry though, I can push to this PR for forever.

haasn commented:
cc @qmega ^ seems you're the de-facto maintainer of the zsh script

occivink commented:
I tested the lavfi-complex changes and it seems to behave as expected regarding the watch_later feature, thanks. Maybe unsetting lavfi-complex could restore the tracks used in the filter?

I also just noticed that the `video-params` properties (in particular, w and h) are after the application of lavfi-complex, is that intended ? But it also used to be the case before these changes.

qmega commented:
2646087567bd6f5c9d8aa719894a6967ad6ac607
Feel free to squash or whatever if you want.

> Alternatively, we could try and come up with a more rigorous (JSON?) method of communicating this data so that the completion script does not need to rely on parsing human-readable output.

This would be nice. Although I'm not sure about JSON; I'd prefer something that I could parse (quickly) without deps using zsh itself so that we could do it at runtime. Anyway, we should have this conversation on #5139 rather than here. Also worth noting that the script in that PR is more flexible and would have handled this change without modification.

kevmitch commented:
Merged  e197ca3dd5d9f5e3f1ea89a35d0cb8de6cb36cf8..d7db42d27f0a474a0ef0de8a771d3e4d5344e3ed
